contracts/SimpleGAMultiSig.aes

```diff
@@ -0,0 +1,153 @@
+// OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
+// PERFORMANCE OF THIS SOFTWARE.
+
+// THIS IS NOT SECURITY AUDITED
+// DO NEVER USE THIS WITHOUT SECURITY AUDIT FIRST
+
+@compiler >= 6
+
+include "Option.aes"
+include "String.aes"
+
+// This contract provides multi signature support based on generalized accounts
+// A transaction can be proposed by one of the multisig users which are registered during contract deployment
+// The transaction must be confirmed by state.confirmations_required of the registered multisig users, while the proposing user automatically confirms
+// If the amount of confirmations is reached, authenticate can be called by any of the multisig users and the original tx is processed
+// Currently only one proposed transaction can be multisigned at a time
+contract SimpleGAMultiSig =
+
+  datatype event
+    = TxProposed(hash, address, int)
+    | TxConfirmed(hash, address)
+    | TxConsensusReached(hash)
+    | TxAuthorized(hash)
+    | TxRevoked(address)
+
+  record transaction = 
+    { ga_tx_hash : hash,
+      expires_at : int,
+      proposed_by : address,
+      confirmed_by : list(address) }
+
+  record state = 
+  	{ multisig_users : list(address),
+  	  confirmations_required : int,    
+      current_tx : option(transaction),
+      nonce : int,
+      owner: address }
+
+  stateful entrypoint init(p_confirmations_required : int, p_multisig_users : list(address)) = 
+    { multisig_users = Call.caller :: p_multisig_users, 
+  	  confirmations_required = p_confirmations_required,
+  	  current_tx = None,
+  	  nonce = 1,
+      owner = Call.caller }
+
+  // GA authorize function
+  stateful entrypoint authorize(sig: signature, n : int) : bool =  
+    check_nonce(n)
+    switch(is_expired())
+      true => 
+        abort("ERROR_TX_EXPIRED")
+      false =>
+        require(has_consensus(), "ERROR_NO_CONSENSUS")
+        switch(Auth.tx_hash)
+          None => abort("ERROR_NO_AUTH_CONTEXT")
+          Some(ga_tx_hash) =>
+            require(ga_tx_hash == Option.force(state.current_tx).ga_tx_hash, "ERROR_UNEQUAL_HASHES")
+            require(is_authorized_and_valid(Call.caller, ga_tx_hash, sig), "ERROR_NOT_AUTHORIZED")
+            Chain.event(TxAuthorized(ga_tx_hash))        
+            true
+    
+  // propose a new tx valid for the given ttl
+  stateful entrypoint propose(ga_tx_hash : hash, ttl : Chain.ttl, sig : signature, n : int) =
+    check_nonce(n)
+    require(is_authorized_and_valid(Call.caller, ga_tx_hash, sig), "ERROR_NOT_AUTHORIZED")
+    switch(state.current_tx)
+      None => set_current_tx(ga_tx_hash, ttl)
+      Some(tx) => 
+        switch(is_expired())
+          false => abort("ERROR_A_TX_IS_ALREADY_PROPOSED")
+          true => set_current_tx(ga_tx_hash, ttl)
+  
+  // signer confirms tx
+  stateful entrypoint confirm(sig : signature, n : int) = 
+    check_nonce(n)
+    switch(state.current_tx)
+      None => abort("ERROR_NOTHING_TO_CONFIRM")
```

Comment:
can be require instead of switch abort

Reply:
but we need the tx here for further checks. what about assigning via `Option.force_msg(state.current_tx, "ERROR_NOTHING_TO_CONFIRM")` to avoid the switch?